include/deal.II/non_matching/quadrature_generator.h

```diff
@@ -0,0 +1,211 @@
+         */
+        double
+        side_length() const;
+
+        /**
+         * Returns the volume of the hypercube.
+         */
+        double
+        volume() const;
+
+        /**
+         * Returns the index:th vertex of the hypercube.
+         */
+        Point<dim>
+        vertex(const unsigned int index) const;
+
+        /**
+         * Returns the index:th child of the hypercube. Child is meant in the
+         * same way as for a cell.
+         */
+        Hypercube<dim>
+        child(const unsigned int index) const;
+
+      private:
+        /**
+         * Coordinate in the center of the hypercube.
+         */
+        const Point<dim> center_coordinate;
+
+        /**
+         * Side length of the hypercube.
+         */
+        const double length_of_side;
+      };
+
+
+      /**
+       * Returns a hypercube between 0 and 1 in all directions: [0, 1]^dim.
+       */
+      template <int dim>
+      Hypercube<dim>
+      create_unit_hypercube();
+
+
+      /**
+       * This function defines a convention for how coordinates in dim
+       * dimensions should translate to the coordinates in dim + 1 dimensions,
+       * when one of the coordinates in dim + 1 dimensions is locked to a given
+       * value.
+       *
+       * The convention is the following: Starting from the locked coordinate we
+       * store the lower dimensional coordinates consecutively and wrapping
+       * around when going over the dimension. This relationship can be
+       * described by the following tables:
+       *
+       *                        2D
+       * ------------------------------------------------
+       * | locked in 2D | 1D coordinate | 2D coordinate |
+       * |--------------------------------------------- |
+       * |     x0       |      (a)      |   (x0,  a)    |
+       * |     x1       |      (a)      |   (a , x1)    |
+       * ------------------------------------------------
+       *
+       *                       3D
+       * --------------------------------------------------
+       * | locked in 3D | 2D coordinates | 3D coordinates |
+       * |----------------------------------------------- |
+       * |     x0       |    (a, b)      | (x0,  a,  b)   |
+       * |     x1       |    (a, b)      | ( b, x1,  a)   |
+       * |     x2       |    (a, b)      | ( a,  b, x2)   |
+       * --------------------------------------------------
+       *
+       * Given a locked coordinate, this function maps a coordinate index in dim
+       * dimension to a coordinate index in dim + 1 dimensions.
+       *
+       * @param locked_coordinate should be in the range [0, dim+1).
+       * @param coordiante_in_dim should be in the range [0, dim).
+       * @return A coordinate index in the range [0, dim+1)
+       */
+      template <int dim>
+      unsigned int
+      coordinate_to_one_dim_higher(const unsigned int locked_coordinate,
+                                   const unsigned int coordiante_in_dim);
```

Comment:
We should make this inline in the header since its just one statement and two assertions.